arcgis-ios-sdk-samples/Utility Network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift

```diff
@@ -0,0 +1,261 @@
+    }
+    
+    func setUIState() {
+        let utilityNetworkIsReady = utilityNetwork.loadStatus == .loaded
+        categoryButton.isEnabled = utilityNetworkIsReady
+        
+        let canTrace = utilityNetworkIsReady && startingLocationElement != nil
+        traceButton.isEnabled = canTrace
+    }
+    
+    /// Clear all the feature selections from previous trace.
+    func clearLayersSelection() {
+        self.mapView.map?.operationalLayers.lazy
+            .compactMap { $0 as? AGSFeatureLayer }
+            .forEach { $0.clearSelection() }
+    }
+    
+    // MARK: - Actions
+    
+    @IBAction func traceButtonTapped(_ button: UIBarButtonItem) {
+        guard let parameters = makeTraceParameters() else { return }
+        SVProgressHUD.show(withStatus: "Running isolation trace…")
+        utilityNetwork.trace(with: parameters) { [weak self] (traceResults, error) in
+            guard let self = self else { return }
+            if let error = error {
+                self.setStatus(message: "Trace failed.")
+                self.presentAlert(error: error)
+                SVProgressHUD.dismiss()
+                return
+            }
+            guard let elementTraceResult = traceResults?.first as? AGSUtilityElementTraceResult, !elementTraceResult.elements.isEmpty else {
+                self.setStatus(message: "Trace completed with no output.")
+                SVProgressHUD.dismiss()
+                return
+            }
+            self.clearLayersSelection()
+            SVProgressHUD.show(withStatus: "Trace completed. Selecting features…")
+            let groupedElements = Dictionary(grouping: elementTraceResult.elements) { $0.networkSource.name }
+            
+            let selectionGroup = DispatchGroup()
+            
+            groupedElements.forEach { (networkName, elements) in
+                guard let layer = self.mapView.map?.operationalLayers.first(where: { ($0 as? AGSFeatureLayer)?.featureTable?.tableName == networkName }) as? AGSFeatureLayer else { return }
+                
+                selectionGroup.enter()
+                self.utilityNetwork.features(for: elements) { [weak self, layer] (features, error) in
+                    defer {
+                        selectionGroup.leave()
+                    }
+                    if let error = error {
+                        self?.setStatus(message: "Selecting features failed.")
+                        self?.presentAlert(error: error)
+                        return
+                    }
+                    guard let features = features else { return }
+                    layer.select(features)
+                }
+            }
+            
+            selectionGroup.notify(queue: .main) { [weak self] in
+                self?.setStatus(message: "Trace completed.")
+                SVProgressHUD.dismiss()
+            }
+        }
+    }
+    
+    @IBAction func categoryButtonTapped(_ button: UIBarButtonItem) {
```

Comment:
As the utility network has fixed presets for this sample, this alert can be created only once and used many times. However, I don't know if it worths to keep a reference to the alertVC.